### ISSUE TITLE
inconsistent port attribute in structs tr_tcp_config and tr_ssh_config

### ISSUE THREAD

**Issue by smlng** [rtrlib/rtrlib#80]
I just stumbled over a weird inconsistency, `tr_tcp_config` is defined as:

```
struct tr_tcp_config {
    char *host;
    char *port;
    char *bindaddr;
};
```

and `tr_ssh_config` as:
```
struct tr_ssh_config {
    char *host;
    unsigned int port;
    char *bindaddr;
    char *username;
    char *server_hostkey_path;
    char *client_privkey_path;
};
```

Is there any reasoning why `port` is a char pointer for tcp and an uint for ssh? IMHO this should be the same (which way ever) in both cases?!

**Comment by waehlisch** [comment]
Good point. As TCP is more frequently used, we should not change the TCP API, I would say. Comments @fho?

**Comment by nbars** [comment]
I think the reason for using a char* is that functions like `getaddrinfo` expect that the port is passed as this type. However I would prefer `unsigned int`, but as @waehlisch said this would break many implementations using the rtrlib.

**Comment by nbars** [comment]
I would resolve this issue when we're fixing #4, because this will anyway change the library API.

**Comment by smlng** [comment]
My question was, why these (very) similar structs have a differing declaration of their port attribute? I wasn't arguing in favor of (or against) `char *` or `unsigned int`, IMO it should be consistent.

Look into [rtrclient.c](https://github.com/rtrlib/rtrlib/blob/master/tools/rtrclient.c#L190), if mode `tcp` port is just passed, and if mode `ssh` its converted into `unsigned int`.

**Comment by shaddeus** [comment]
TCP transport `getaddrinfo()` [tcp_transport.c:53](https://github.com/rtrlib/rtrlib/blob/master/rtrlib/transport/tcp/tcp_transport.c#L53) needs representation in `const char *`. SSH transport [ssh_transport.c:55](https://github.com/rtrlib/rtrlib/blob/master/rtrlib/transport/ssh/ssh_transport.c#L55) expects port number as 'unsigned int'. It is better to convert string to `unsignet int` only once at the start than at each connection attempt.

**Comment by smlng** [comment]
@shaddeus: ah I see, thanks for the pointer. 

So these differing types make sense from a library (internal) perspective, but might be a bit confusing for an external developer using these structs. 

So maybe @nbars is right and we resolve this issue via fixing #4, i.e., hide such internal structs from the developer/application.

**Comment by waehlisch** [comment]
@shaddeus as a user of RTRlib, what do you think about #4?

**Comment by shaddeus** [comment]
@waehlisch thanks for a call!

The RTRlib is multithreaded stand-alone living library and in my opinion it would be nice to **make a deep copy of all parameters** passed into RTRlib and put them into `struct rtr_mgr_config` in `rtr_mgr_init()`

Anyway:

- move a triple `const char *host, *port, *bindaddr` into common `struct tr_socket` and then make a deep copy 

- remove public `struct tr_tcp_config` becouse it would be empty.

- either to remove `struct tr_ssh_config` too and pass parametrs into `tr_ssh_init(tr_socket, ...)` or to let `struct tr_ssh_config` but ensure a deep copy

**Comment by fho** [comment]
Like @nbars mentioned, the TCP socket supports to pass a service name (/etc/services) or a port number as char *. It's then resolved from getaddrinfo to a numeric value.
The SSH socket only supports numeric port numbers.

Yes, would be better if it would be the same type for both.

Solving #4 would be the best solution. It would also allow to change the internal representation without effecting the external API.

**Comment by shaddeus** [comment]
The SSH socket could be set by `const char *` too using `SSH_OPTIONS_PORT_STR` instead of `SSH_OPTIONS_PORT`. See [LibSSH API: Session: ssh_options_set()](http://api.libssh.org/master/group__libssh__session.html#ga7a801b85800baa3f4e16f5b47db0a73d)
